### PR TITLE
[extractor/vk] playlist fix for new API requirements

### PR DESCRIPTION
**Pull request by the-marenga**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

As reported [here](https://github.com/yt-dlp/yt-dlp/pull/4930#issuecomment-1409576726), VK playlist downloads stopped working. The reason is, that their API has been changed yet again and now requires "playlist_ID" instead of just "ID" as a request parameter. With this fix everything works again

Fixes  https://github.com/yt-dlp/yt-dlp/pull/4930#issuecomment-1409576726

# PLEASE FOLLOW THE GUIDE BELOW

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/vk.py [hunk 1]
```diff
@@ -595,6 +595,7 @@ def _real_extract(self, url):
             page_id = self._search_regex(r'data-owner-id\s?=\s?"([^"]+)"', webpage, 'page_id')
         elif '_' in u_id:
             page_id, section = u_id.split('_', 1)
+            section = 'playlist_' + section
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:598]:
```suggestion
            section = f'playlist_{section}'
```

**Comment by the-marenga** [comment]
Yes, the other branches are for URLs, that either provide their own valid section id, or download an entire channel 'all'. Both of these are still what VK uses in their requests unchanged.

I did however just notice, that I forgot to update the playlist test, so I fixed that. 

When doing so I also noticed, that most of the single video (VKIE) tests either fail for various reasons, or get skipped because the video has been removed. I think they were already failing months ago, but I am not sure. If you want me to, I can just update all of the tests in the extractor (either in this PR, or another one?)  to make sure nothing there has been broken by this or an older API change

**Comment by pukkandan** [comment]
> If you want me to, I can just update all of the tests in the extractor (either in this PR, or another one?) to make sure nothing there has been broken by this or an older API change

Sure, you can do it in same PR

**Comment by the-marenga** [comment]
I updated all the tests and it seems like they did in fact change their site somewhere else.

They changed their wallposts audio data to JSON and removed the url encryption, which crashed the previously already non-functioning audio extraction completely. I fixed that, but downloading still returned an error from the 'FFmpegFixupM3u8PP', because the audio stream is mp3 and not the expected aac. I changed the container to mov to hide the error for now. The files would however actually need a simple fixup, as they are currently only playable with a handful of media players. A real fix for this would be to have a url check in the fixup, to ommit the incompatible stuff:
```python
if all(self._needs_fixup(info)):
    if 'vkuseraudio.net' in info['url']:
        self._fixup('Fixing MPEG-TS in MP4 container', info['filepath'], [
            *self.stream_copy_opts(), '-f', 'mp4'])
    else:
        self._fixup('Fixing MPEG-TS in MP4 container', info['filepath'], [
            *self.stream_copy_opts(), '-f', 'mp4', '-bsf:a', 'aac_adtstoasc'])
```
but that seems like an inelegant way to do this and I am not familiar enough with the codebase to come up with a better alternative. For now you would either have to play the files in something like mpv, or use -x to extract the actual audio manually.

VK also lifted a lot of restrictions to being registered, so a lot of tests for edge cases are no longer required and some don't need to be skipped anymore.

All tests succeed now

File: yt_dlp/extractor/vk.py [hunk 2]
```diff
@@ -710,21 +672,20 @@ def _real_extract(self, url):
...
-                'thumbnails': [{'url': c_url} for c_url in a.cover_url.split(',')] if a.cover_url else None,
-                'duration': int_or_none(a.duration),
+                'id': '%s_%s' % (audio['owner_id'], audio['id']),
+                'url': audio['url'],
+                'title': '%s - %s' % (artist, title) if artist else title,
+                'thumbnails': [{'url': c_url} for c_url in audio['coverUrl'].split(',')] if audio['coverUrl'] else None,
+                'duration': int_or_none(audio['duration']),
                 'uploader': uploader,
-                'artist': performer,
+                'artist': artist,
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:686]:
```suggestion
            entries.append({
                'id': f'{audio["owner_id"]}_{audio["id"]},
                'url': audio['url'],
                'title': join_nonempty(audio.get('artist'), audio.get('title'), delim=' - '),
                'thumbnails': try_call(lambda: [{'url': u} for u in audio['coverUrl'].split(',')]),
                'duration': int_or_none(audio.get('duration')),
                'uploader': uploader,
                'artist': audio.get('artist'),
```

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
> The files would however actually need a simple fixup, as they are currently only playable with a handful of media players.

If it's mp3, what kind of fixup does it need?

File: yt_dlp/extractor/vk.py [hunk 3]
```diff
@@ -710,21 +672,20 @@ def _real_extract(self, url):
...
+            title = unescapeHTML(audio['title'])
+            artist = unescapeHTML(audio['artist'])
             entries.append({
-                'id': '%s_%s' % (a.owner_id, a.id),
-                'url': self._unmask_url(a.url, a.ads['vk_id']),
-                'title': '%s - %s' % (performer, title) if performer else title,
-                'thumbnails': [{'url': c_url} for c_url in a.cover_url.split(',')] if a.cover_url else None,
-                'duration': int_or_none(a.duration),
+                'id': '%s_%s' % (audio['owner_id'], audio['id']),
+                'url': audio['url'],
```
Comment (pukkandan) [inline yt_dlp/extractor/vk.py:681]: I think I found the issue. Try removing `url`, `ext`, `protocol` and add `'formats': self._extract_m3u8_formats(...)` instead

File: yt_dlp/extractor/vk.py [hunk 4]
```diff
@@ -117,7 +117,7 @@ class VKIE(VKBaseIE):
                 'upload_date': '20120212',
                 'comment_count': int,
                 'like_count': int,
-                'thumbnail': r're:https?://.+\.jpg$',
+                'thumbnail': r're:https?://.+(?:\.jpg|getVideoPreview.*)$',
```
[the remark this answers is not in the document]
Comment (the-marenga) [reply yt_dlp/extractor/vk.py:120]: Yes, the url can now be of the form `https://i.mycdn.me/getVideoPreview?id=888771774995&idx=3&type=39&tkn=cCEg28A6lldfIr7gna_Xz1wqKtc&fn=vid_x`, which contains a webp. When I use `--write-thumbnail`, I get a jpg containing the image. I dont know if you convert the webp -> jpg somewhere, or if they serve another format based on the request headers (Accept/Useragent) yt-dlp sends, but it overall seems to work. 
I only saw this change for one of the tests, but it seems like something they would extend to all thumbnails over time, so I updated all tests for it

File: yt_dlp/extractor/vk.py (hunk 3 above)
Comment (the-marenga) [reply yt_dlp/extractor/vk.py]: I thought so too, because it works for vk videos, but the files this outputs are no different to the ones before. It just doesn't seem to find anything:
```
[info] Available formats for 2000326341_456244816:
ID EXT RESOLUTION │ PROTO │ VCODEC  ACODEC
───────────────────────────────────────────
0  mov unknown    │ m3u8  │ unknown unknown
```
 I tried it with live=true/false and m3u8_id=None/'hls'. Ffprobe identifies the resulting files as:
```
mpegts @ 0x5614dee86000] start time for stream 1 is not set in estimate_timings_from_pts
Input #0, mpegts, from 'input.mp4':
  Duration: 00:06:10.44, start: 0.127122, bitrate: 352 kb/s
  Program 1 
  Stream #0:0[0x100]: Audio: mp3 ([3][0][0][0] / 0x0003), 44100 Hz, stereo, fltp, 320 kb/s
  Stream #0:1[0x101]: Data: timed_id3 (ID3  / 0x20334449)
Unsupported codec with id 98313 for input stream 1
```
This will not be playable with  VLC, which would be my standard for "working" media files. MPV however manages to play that. If you fix them with `ffmpeg -i input.mp4  -map 0 -dn -ignore_unknown -c copy -f mp4 output.mp4` they become:
```
Input #0, mov,mp4,m4a,3gp,3g2,mj2, from 'output.mp4':
  Metadata:
    major_brand     : isom
    minor_version   : 512
    compatible_brands: isomiso2mp41
    encoder         : Lavf59.27.100
  Duration: 00:06:10.44, start: 0.000000, bitrate: 321 kb/s
  Stream #0:0[0x1](und): Audio: mp3 (mp4a / 0x6134706D), 44100 Hz, stereo, fltp, 320 kb/s (default)
    Metadata:
      handler_name    : SoundHandler
      vendor_id       : [0][0][0][0]
```
These are now playable without any issues everywhere. Given that m3u8 already has a fixup for broken audio (that just doesnt work here because of mp3 instead of aac audio) I would guess, that this is a somewhat known problem, that a user is not expected to solve on their own?

Comment (pukkandan) [reply yt_dlp/extractor/vk.py]: Try inserting `{'vcodec': 'none', 'acodec': 'mp3', 'container': 'm4a_dash'}` into the format

**Comment by the-marenga** [comment]
Great, the files are playable and have no corruptions. 
I just changed the author extraction class, because vk changed it within the last few days and updated the tests.
Only thing left would be merging the audio & video entries again when returning, but I couldn't figure out how to append the audio entries to the `playlist_from_matches` without basically reverting that change from your cleanup back to using 'playlist_result'

**Comment by pukkandan** [comment]
> Only thing left would be merging the audio & video entries again when returning, but I couldn't figure out how to append the audio entries to the `playlist_from_matches` without basically reverting that change from your cleanup back to using 'playlist_result'

My bad. Reverted

**Comment by the-marenga** [comment]
When doing final testing, I noticed, that irrelevant reaction clips in the comments will also be picked up. That seemed like a bug to me, so I fixed that by only searching for videos in the actual post body. Otherwise everything should be good
